src/jdk.compiler/share/classes/com/sun/tools/javac/code/LintMapper.java

```diff
@@ -0,0 +1,329 @@
+        final List<Span> unmappedDecls = new ArrayList<>();     // unmapped top-level declarations awaiting attribution
+
+        // After parsing: Add top-level declarations to our "unmappedDecls" list
+        FileInfo(Lint rootLint, JCCompilationUnit tree) {
+            rootRange = new LintRange(rootLint);
+            tree.defs.stream()
+              .filter(this::isTopLevelDecl)
+              .map(decl -> new Span(decl, tree.endPositions))
+              .forEach(unmappedDecls::add);
+        }
+
+        // After attribution: Discard the span from "unmappedDecls" and populate the declaration's subtree under "rootRange"
+        void afterAttr(JCTree tree, EndPosTable endPositions) {
+            for (Iterator<Span> i = unmappedDecls.iterator(); i.hasNext(); ) {
+                if (i.next().contains(tree.pos())) {
+                    rootRange.populateSubtree(tree, endPositions);
+                    i.remove();
+                    return;
+                }
+            }
+            throw new AssertionError("top-level declaration not found");
+        }
+
+        // Find the most specific Lint configuration applying to the given position, unless the position has not been mapped yet
+        Optional<Lint> lintAt(DiagnosticPosition pos) {
+            boolean mapped = unmappedDecls.stream().noneMatch(span -> span.contains(pos));
+            return mapped ? Optional.of(rootRange.bestMatch(pos).lint) : Optional.empty();
+        }
+
+        boolean isTopLevelDecl(JCTree tree) {
+            return tree.getTag() == Tag.MODULEDEF
+                || tree.getTag() == Tag.PACKAGEDEF
+                || tree.getTag() == Tag.CLASSDEF;
+        }
+    }
+
+// Span
+
+    /**
+     * A lexical range.
+     */
+    private record Span(int startPos, int endPos) {
+
+        static final Span MAXIMAL = new Span(Integer.MIN_VALUE, Integer.MAX_VALUE);
+
+        Span(JCTree tree, EndPosTable endPositions) {
+            this(TreeInfo.getStartPos(tree), TreeInfo.getEndPos(tree, endPositions));
+        }
+
+        boolean contains(DiagnosticPosition pos) {
+            int offset = pos.getLintPosition();
+            return offset == startPos || (offset > startPos && offset < endPos);
+        }
+
+        boolean contains(Span that) {
+            return this.startPos <= that.startPos && this.endPos >= that.endPos;
+        }
+    }
+
+// LintRange
+
+    /**
+     * A tree of nested lexical ranges and the {@link Lint} configurations that apply therein.
+     */
+    private record LintRange(
+        Span span,                                      // declaration's lexical range
+        Lint lint,                                      // the Lint configuration that applies at this declaration
+        Symbol symbol,                                  // declaration symbol (for debug purposes only; null for root)
+        List<LintRange> children                        // the nested declarations one level below this node
```

Comment:
not a proposal to do any change, but I think that if instead of having a tree-like structure in which each range has, potentially, a number of children, we had one list of ranges sorted by its starting point. I think that we could find the right range doing a binary search. We could actually have both: the current tree and a flat list of lint ranges for searches. Again probably not necessary for general cases but there could be corner cases as in machine generated code for which having a faster retrieval could make the difference.

Reply:
I think at one point I had something like that, but it made things more complicated. Plain binary search doesn't work when you are searching in a list of (possibly nested) intervals instead of a list of points, because there can be multiple matches (arbitrarily many in fact). I think it would require an [Interval tree](https://en.wikipedia.org/wiki/Interval_tree).

There is also another annoying detail, which is that the nesting is not always proper: in some oddball cases (none of which I can specifically remember) there can be an AST node whose starting position is before the starting position of its parent AST node (or ending position after its parent). These instances could can an Interval tree type of algorithm to glitch out.